### DISCUSSION TITLE
Expand only at beginning of words

### DISCUSSION THREAD

**Discussion by SmilingGit** [espanso/espanso#1653] [Feature requests and Ideas]
Hello,

Thanks again for this great tool, it’s a joy to use. This request has some similarities to this thread: https://github.com/espanso/espanso/issues/1411

In the Windows-World, I had been using Autohotkey, and somehow a standard setting of Autohotkey is to start triggers only at the beginning of words (you can deactivate that behaviour of course). I’ve been looking for that feature everywhere in Espanso, and I believe it’s not there.

Usecase: I would like immediately expanding triggers, that don’t trigger inside of a word. For instance, I’d like to have a trigger "fk" that triggers only at the beginning of a word. There’s no word I know of in a language I use that starts with "fk". However, there are words in German that have "fk" inside a word. So when I start a word and type "fk", I want it to expand immediately, but only then. 

There might be a way of doing this with regex, but I haven’t found one, and I also think it creates a steep learning curve for newcomers.

[Just to be complete, I’ve tried triggers starting with ^ (works only at beginning of text, even with (?m), and doesn’t capture white spaces), starting with \b (doesn’t trigger at all), with [[:space:]] (triggers, but doesn’t trigger at the beginning of a new text), \s (creates an error while "compiling", although it’s in the documentation of the Rust-Regex-library). Maybe I could use several lines of triggers, but then it would become really tedious.] [Edit: found why the \-triggers didn’t work, because I didn’t double them. Putting \\b at the beginning of the word actually does the trick. That’s great. I still think an option would be great. An in the meantime, maybe someone will find this discussion. Or a paragraph can be added to the documentation?]

How I would implement it: a bit like in the post mentioned above. Adding one or several options to the "word"-option (or adding another option).
Like: word_beginning: true
Or word: beginning

That’s it, I hope I have been clear! 

Have a great day!

**Comment by jwigg-netscout** [comment]
sorry, isn't this just [word: true](https://espanso.org/docs/matches/basics/#word-triggers)? is there some nuance I'm missing?

**Comment by SmilingGit** [comment]
With word:true, you need to begin and end a word for it to be expanded. I, however, would like it to expand as soon as I typed the trigger, not wait for a space or something else.
Example: for it to trigger with the abovementioned  fk I need to type, for instance, space,  fk, then space. I would like it to trigger immediately after the k. But I don’t want it to trigger if I type ffk for instance, which happens when I don’t activate word:true. Is that clearer now? 
The workaround with regex and \\b works OK, but I still think it would be more practical as a full feature. I guess there must be a good reason Autohotkey has been working like this for so long.

**Comment by jwigg-netscout** [comment]
Ok, I think I understand now. At the moment, I can't see myself using that feature, but I agree that if that's the behavior you want then having to specify the regex that way for each match would be onerous 🙁

**Comment by SmilingGit** [comment]
By the way, I tried changing many of my triggers to regex, but it makes it extremely complicated, since the preserve case option doesn’t work any more, and I’d have to handle the cases manually for it to work…

**Comment by jwigg-netscout** [comment]
is the choice of "fk" just to have an unambiguous 2 character prefix that you can start matches with? If so, can I suggest the prefix I use: ";;" ? It has the following properties:

- available on the home row of many keyboard layouts, across multiple languages and variants (US, UK, CN, etc)
  - still relatively accessible in most other keyboard layouts (DE, dvorak, colemak, etc)
- not valid punctuation in any language I know of
- not a valid operator (as a prefix) in any programing language I know of

In other words, ;;trigger isn't something I'm ever likely to type on purpose, in any language or code, unless I'm intentionally trying to trigger a replacement. I've been using it for over a decade and have yet to encounter a situation in English, German, Python, Bash, Javascript, PHP, or Ruby where it has ever accidentally triggered a replacement. 

If the choice of "fk" is for some other reason, maybe if you explain it we can find a solution?

**Reply by SmilingGit** [reply]
Thanks for your answer!

"fk" is just an example of many many triggers. I like to have triggers without prefix, that makes them faster to type. And so I usually choose triggers that are never at the beginning of a word. However, if I have to pick triggers that are neither at the beginning of a word, nor in the middle, it limits my option by a lot. 

As a sidenote, I’ve reconfigured my keyboard quite heavily: as a result, I don’t have rarely used characters easily accessible. And to make them easier to remember, I don’t have many triggers that combine letters and special characters. Thank you anyway for the suggestion, If there’s no other possible option, I’ll try so see if I can play with more special characters for my triggers.
